Add solidity wrappers for CIP-20/30/31 precompiles

karlb opened the pull request:
### Description

CIP-20/30/31 add new precompiles to the celo blockchain. Using these is hard without a set of solidity wrappers. This PR provides these wrappers.

Original implementation from https://github.com/prestwich/b12-sol.

I'm not sure if it was correct to use `forge install` to install the memview package. If it was, I wonder how I should proceed with the [tslint error](https://app.circleci.com/pipelines/github/celo-org/celo-monorepo/44172/workflows/fb258e07-ea09-47ed-aca4-536d9816df44/jobs/584188/parallel-runs/0/steps/0-103). Manually editing external dependencies to fix lint errors seems wrong, since it will make it harder to update dependencies.

### Tested

Since the wrappers can only be tested against a blockchain implementation with support for the respective precompiles, we can't use ganache to run the tests. Therefore, the wrappers are tested in the celo-blockchain repo against the real implementation.

### Related issues

- Implementation for https://github.com/celo-org/celo-blockchain-planning/issues/42
- Testing happens in https://github.com/celo-org/celo-blockchain/pull/2058

### Backwards compatibility

No existing contracts are changed -> backwards compatible

### Documentation

See https://github.com/celo-org/docs/pull/872

karlb committed:
Add B12 precompiles (CIP-30/31) and fix style

karlb committed:
Add CIP20Lib and fix style errors

karlb committed:
Add test contracts for CIP-20, CIP-30 and CIP-31

karlb committed:
Use B12_377Lib in BLS12Passthrough

The CeloB12_377Lib has precompile address which match neither our current
implementation nor CIP30.

karlb committed:
Remove CeloB12_377Lib

The CeloB12_377Lib has precompile address which match neither our current
implementation nor CIP30. The same precompile wrapper implementation
with the correct precompile addresses is available in B12_377Lib.

karlb committed:
Use same solidity pragma and license for sol files

The code allows Apache-2 licensing. Kept link to orig repo.

karlb committed:
Make forge work with new precompile wrappers

* Add dependency
* Ignore passthrough test (tested from celo-blockchain repo)

m-chrzan commented on packages/protocol/contracts/common/B12.sol:
```diff
@@ -0,0 +1,898 @@
+  function fpSub(Fp memory a, Fp memory b) internal pure returns (Fp memory) {
+    Fp memory p = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
```
Could these magic numbers be pulled out as some meaningfully named `constant`s? Same for other magic numbers throughout.

karlb replied:
I'm afraid I lack the context to find meaningful names for these.

m-chrzan commented on packages/protocol/contracts/common/B12.sol:
```diff
@@ -0,0 +1,898 @@
+      mstore(add(arg, 0x100), modulus2)
+
+      // call the precompiled contract BigModExp (0x05)
+      let success := staticcall(gas, 0x05, add(arg, 0x20), 0x100, add(ret, 0x20), 0x40)
```
Would be nice to not use a magic number here for the precompile address.

Ideal scenario is we have a library for using the new Byzantium precompiles (does an external library exist?) that are not built into Solidity.

Can be out of scope for this PR, but would create an issue to deal with this later.

karlb replied:
I did not find an external library. We could pull out all precompile addresses (both Eth and Celo) into a precompile address lib and use that in UsingPrecompiles.sol, B12.sol and CIP20Lib.sol. But then these files would not be self-contained anymore, which might be nice for users including these files in their project (since we don't provide them as library packages). I'm not sure that would be an advantage overall.

I can certainly pull out this magic number to a variable, but since the this is the only Eth precompile we use here, that also would not be clearly better than this comment.

I don't have a strong opinion on any of the approaches.

pahor167 replied:
Since this file seems to be almost copy of https://github.com/prestwich/b12-sol/blob/main/contracts/B12.sol I would personally vote for leaving it close to original source as possible.

m-chrzan commented on packages/protocol/contracts/common/B12.sol:
```diff
@@ -0,0 +1,898 @@
+
+  function parseSimplePoint(bytes memory h, uint256 offset, Fp memory p) internal pure {
+    /* (uint256 a, uint256 b, uint256 byt) = parsePointGen(h, offset);
+    a = a + (byt << 15*8); */
```
Not sure if these are explanatory comments or unused code. Seems more like the latter? If that's the case, do we know for sure which is correct, if it's safe to remove the commented out code?

karlb replied:
Looks unused to me, too. I'll remove it, even though I don't know for sure.

m-chrzan commented on packages/protocol/contracts/common/CIP20Lib.sol:
```diff
@@ -0,0 +1,140 @@
+      mstore(input, len)
+    }
+
+    require(success, "keccak512 failed");
```
Does this only fail when keccak512 is selected? Or should this be more general, like "CIP20 hashing failed"?

karlb replied:
You are right. Fixed.

m-chrzan commented on packages/protocol/contracts/common/CIP20Lib.sol:
```diff
@@ -0,0 +1,140 @@
+  }
+
+  function sha3_256(bytes memory input) internal view returns (bytes memory) {
+    return executeCip20(input, SHA3_256_SELECTOR, 32);
```
What do you think about defining the output lengths as `constant`s as well, to avoid magic numberage?

karlb replied:
I think in this case it is obvious enough that 256BIT_LENGTH_IN_BYTES would not be an improvement.

m-chrzan commented on packages/protocol/contracts/common/CIP20Lib.sol:
```diff
@@ -0,0 +1,140 @@
+  uint8 private constant SHA2_512_SELECTOR = 0x03;
+  uint8 private constant BLAKE2S_SELECTOR = 0x10;
+
+  bytes32 private constant BLAKE2S_DEFAULT_CONFIG = 0x2000010100000000000000000000000000000000000000000000000000000000;
```
Would it be useful to comment what this config represents (e.g. a comment with a call to `createConfig` with the appropriate arguments that would result in this config, assuming that's a way to get this)?

karlb replied:
Yes that can be created with createConfig and I even found the right values. Fixed.

m-chrzan commented on packages/protocol/contracts/common/CIP20Lib.sol:
```diff
@@ -0,0 +1,140 @@
+    return blake2sWithConfig(BLAKE2S_DEFAULT_CONFIG, hex"", preimage);
+  }
+
+  function createConfig(
```
Assuming this is for Blake2s? Should the name be more explicit, something like `createBlake2sConfig`?

karlb replied:
Agreed and done.

m-chrzan commented on packages/protocol/contracts/common/test/BLS12Passthrough.sol:
```diff
@@ -0,0 +1,354 @@
+    require(success, "failed");
+  }
+
+  function simple(bytes calldata input, uint8 addr, uint256 output_len)
```
Is this a descriptive name for this function? Is it even necessary if it does the same exact thing as `executePrecompile`?
My suggestion would be to remove this, make `exeuctePrecompile` `public`, rename `simpleTx` to `executePrecompileWithOutput` (or similar).

karlb replied:
They are not even run as part of the tests. They seem to be dead code, so I'll just remove them.

m-chrzan commented on packages/protocol/contracts/common/test/BLS12Passthrough.sol:
```diff
@@ -0,0 +1,354 @@
+  using TypedMemView for bytes29;
+
+  constructor() public {}
+  function testParseG1(bytes calldata arg) external pure returns (uint256[4] memory ret) {
```
Are the "test" pre-/suffixes necessary throughout this file?

karlb replied:
They don't serve any technical purpose, but they make it clear that this contract is not meant for reuse and that this specific function will be called from outside as part of a test run.

m-chrzan commented on packages/protocol/lib/memview-sol/.gitignore:
```diff
@@ -0,0 +1,6 @@
+node_modules/
```
Are the files in `lib/memview-sol/` all from the `foundry` installation? Is it typical to commit these to a repo using a `foundry` package, or should this be all ignored?

karlb replied:
> Are the files in lib/memview-sol/ all from the foundry installation

Yes.

> Is it typical to commit these to a repo using a foundry package, or should this be all ignored?

I have no idea. I only discovered foundry during the CI failure and have never used it before. I would have preferred to not commit these files, but since I couldn't find an existing build step where new dependencies are installed this seemed like the least invasive way to make it work.

I could also add more foundry ignores so that this is not even necessary, since the code is tested with the TS test runner.

As mentioned in the PR description I was hoping for some advice on the foundry situation:
> I'm not sure if it was correct to use forge install to install the memview package. If it was, I wonder how I should proceed with the [tslint error](https://app.circleci.com/pipelines/github/celo-org/celo-monorepo/44172/workflows/fb258e07-ea09-47ed-aca4-536d9816df44/jobs/584188/parallel-runs/0/steps/0-103). Manually editing external dependencies to fix lint errors seems wrong, since it will make it harder to update dependencies.

m-chrzan replied:
Looking at some of the other foundry packages (openzeppelin-contracts, solidity-bytes-utils, etc.), their directories in the repo are empty.

> but since I couldn't find an existing build step where new dependencies are installed this seemed like the least invasive way to make it work

Yep, also not seeing how this happens. Not very familiar with foundry myself though. Are there missing steps from our build scripts that should be added? Do you know how/when the other foundry packages are installed (like openzeppelin-contracts)?

That said, to your point of whether we should even be using foundry for this, I think there is an argument to be made for just copying the files over to our repo and maintaining them ourselves rather than relying on an external package.

karlb replied:
@bowd You added the foundry usage. Can you give us some hints on this?

bowd replied:
Late to the party: 

So the `lib` folder in foundry projects is reserved for 3rd party dependencies which are linked as submodules. 
That's why you see those folders as empty. You need to run `forge install` or `git submodule update` to pull in the deps.

So it's not usual practice to actually commit the full code in this repo. However I see you mentioned you used `forge install` to install it, but that command should pull it as a submodule, did you then manually remove the submodule and committed the code instead?

karlb replied:
Thanks for the info. I did not intentionally remove the submodule, but maybe I accidentally did when pulling the changes apart into separate commits. Not that I know what the result should look like, I can redo it to get the desired result.

m-chrzan commented:
Oh, one more comment. A more appropriate place for the libraries could be under `contracts/common/libraries/`.

karlb commented:
> Oh, one more comment. A more appropriate place for the libraries could be under contracts/common/libraries/.

Make sense. I just took UsingPrecompiles as an example, since it serves a similar purpose.

karlb committed:
Fix error message in executeCip20

Apparently this has not been updated after making the function generic.

karlb committed:
Explain default blake2 config in comment

karlb committed:
Fix typo

karlb committed:
Rename createConfig to createBlake2sConfig

karlb committed:
Remove unused `simple*` functions in BLS12Passthrough

karlb committed:
B12.sol: Remove commented code (looks obsolete)

karlb committed:
Move CIP20/30/31 prec-wrappers to common/libraries

m-chrzan commented:
> > Oh, one more comment. A more appropriate place for the libraries could be under contracts/common/libraries/.
> 
> Make sense. I just took UsingPrecompiles as an example, since it serves a similar purpose.

Ah, makes sense. Actually, looking at the `libraries/` directory, it doesn't look like we're really using it nor what it's exact purpose is - there's one library and one *contract* there, that could instead probably live just in `common/` given that's where all the other libraries and helper contracts live right now. Given that, I'd keep your libraries where you put them for now, maybe a future task to reorganize the file structure of `contracts/`more consistently/usefully.

karlb committed:
Fix wrong lib usage in BLS12Passthrough

karlb reviewed (commented):
Please note that I am not the original author of the code, so I also just deduct my opinions from the code I see and can't speak about the original intentions.

m-chrzan commented:
> Please note that I am not the original author of the code, so I also just deduct my opinions from the code I see and can't speak about the original intentions.

Yep, definitely looking at all this with that context in mind.

That said, if we're including this in our codebase and taking over maintenance, there's definitely some polishing that can/should be done. Especially since, at least to me, it's not clear how finazlied/polished the original code is.

One thing in particular that would be very useful is NatSpec documentation for all the functions, as we conventionally include them for all our Solidity code. Not sure how much cryptography expertise you have and how comfortable you'd be with filling these out (some of them should be quite simple, others might require deeper understanding), or if we should reach out to our crypto experts for help with that.

karlb commented:
> Not sure how much cryptography expertise

I'm just a developer, not a cryptographer. I can read the code but I don't know the math behind it. So I don't think it makes sense for me to add documentation on the actual crypto part. There's not that much more on that front that I can provide beyond the basic introduction I wrote at https://github.com/celo-org/docs/pull/872.

m-chrzan commented:
> So I don't think it makes sense for me to add documentation on the actual crypto part. There's not that much more on that front that I can provide beyond the basic introduction I wrote at [celo-org/docs#872](https://github.com/celo-org/docs/pull/872).

Sounds good, thanks! [Crated an issue](https://github.com/celo-org/celo-monorepo/issues/10242) to track this. Maybe someone could take the low hanging fruit of the more obvious functions (maybe me if I find the time), then would be good to have someone like @kobigurk  fill in the rest.

m-chrzan commented on packages/protocol/contracts/common/libraries/B12.sol:
```diff
@@ -0,0 +1,895 @@
+    Fp memory p = Fp(
+      0x1ae3a4617c510eac63b05c06ca1493b,
+      0x1a22d9f300f5138f1ef3622fba094800170b5d44300000008508c00000000001
+    );
```
Don't have the expertise to say for sure, but just want to say this looks fishy to me and it'd be great to get someone that knows more about BLS crypto to look at this.

The number here appears to be the base field modulus for BLS377 (see [here](https://eips.ethereum.org/EIPS/eip-2539#specification) for example; different from the BLS381 `p` as seen e.g. [here](https://github.com/zkcrypto/bls12_381)).

It doesn't appear that this function ever (directly or indirectly) gets called by the BLS381 library, but this just points to some weird organization of these libraries. Seems like some of the BLS377-specific functions are defined in B12, some in the BLS377 library, while there doesn't seem to be any BLS381-specific functions in B12. It might be that the code is still all correct, but I'm afraid this sort of code organization could lead to future developer confusion (for cLabs or external devs relying on these wrappers).

m-chrzan replied:
Looking a little more deeply, seems like all the `fp<operation>` functions are only used to implement `mapToG1` and `mapToG2` in B12, in an apparently BLS377-specific way. The BLS381 library implements these `mapTo` operations calling out to precompiles. [CIP30](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0030.md) (for BLS-377) did not add these "map to G1/2" precompiles, while [CIP31](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0031.md), for BLS-381, did.

Since the precompiles are not available for BLS-377, makes sense to implement them manually in Solidity, but I guess it would make more sense for these functions to exist in the BLS377 library, rather than the general B12.

karlb replied:
Some fp functions are used in BLS12Passthrough.sol directly from `B12` and not through one of the specialized libraries. I'm not sure if such a generic usage is a relevant use case outside of testing, but it is at least one reason against moving the fp functions.

karlb commented:
Is there anything I can do on my side make this ready for merging?

eelanagaraj commented on packages/protocol/contracts/common/libraries/CIP20Lib.sol:
```diff
@@ -0,0 +1,141 @@
+
+  // Accepts a fully formed input blob. This should include any config
+  // options and the preimage, but not the selector.
+  function executeCip20(bytes memory input, uint8 selector, uint256 output_len)
```
nit: `Cip` -> `CIP` in naming throughout for consistency

karlb replied:
Done here and in `common/libraries/test/Cip20Test.sol`, where I also changed the file name to the consistent casing.

eelanagaraj commented on packages/protocol/contracts/common/libraries/test/Cip20Test.sol:
```diff
@@ -0,0 +1,36 @@
+
+import "../CIP20Lib.sol";
+
+contract Cip20Test {
```
nit: same naming nit, can we make this `CIP20Test`?

eelanagaraj reviewed (commented):
While I can make sense of most of the CIP-20 work, I feel really unqualified to properly review anything pertaining to the BLS implementations. At minimum we should have someone from the crypto team review this work and let us know what additional steps/checks should be done before merging this + the related tests in the blockchain repo IMO.

From what I could tell/check, all config numbers/params in `CIP20Lib.sol` match those listed in the CIP-20 spec and I couldn't find anything wrong with any of the memory offsets.

karlb committed:
Rename Cip20 -> CIP20 for consistency

karlb committed:
Remove memview-sol from repo to allow submodule

The lib folder in foundry projects is reserved for 3rd party dependencies which are linked as submodules.

karlb committed:
Add memview.sol as a submodule

using `forge install`

karlb committed:
Fix solhint errors

montera82 commented on packages/protocol/contracts/common/libraries/test/CIP20Test.sol:
```diff
@@ -0,0 +1,36 @@
+    return CIP20Lib.blake2sWithConfig(config, key, preimage);
+  }
+
+  // default settings, no key
```
nit. do we really need this comment?

karlb replied:
We don't. I don't mind much either way, so I'll just remove it in both places, since you seem to have a slight preference that way.

I don't think we to rename `blake2s`, though. Having a short name with the default settings seems fine to me.

montera82 commented on packages/protocol/contracts/common/libraries/CIP20Lib.sol:
```diff
@@ -0,0 +1,141 @@
+    return executeCIP20(configuredInput, BLAKE2S_SELECTOR, uint256(uint8(config[0])));
+  }
+
+  // default settings, no key
```
same. The comments didn't really help me - perharps we could had some more explanation of what you mean, otherwise we can check if it is actually needed. Great work btw

montera82 replied:
or we can rename function to `blake2sWithoutConfig` 
seems verbose howerver the entire code could help with further handholding cos i'm not too familiar with whats going on mostly anyways 🤷

montera82 commented:
Firstly welldone @karlb  for such a cryptographic PR, not sure whats mostly going on here as well cos my knowledge of cryptography null.
However i noticed the existing file for `usingPrecompiles.sol` has got natspec documentations. But i realise it'll be a tough to document most of the ones in this PR since the original author didn't include much. 
Only left a few nits . Cheers.

karlb committed:
Remove "default settings" comment

karlb committed:
B12: Extract BLS12_377_BASE

martinvol commented:
Regarding the failed Google Cloud Build (docker images). I think it is because submodules are not installed with `yarn build`, so the images can't get built. I already fixed this here: https://github.com/celo-org/celo-monorepo/pull/10124

So when my PR gets it should turn green without any conflicts.

Also there's one image that should be deprecated, so will remove it first in my PR.

pahor167 commented on packages/protocol/contracts/common/libraries/B12.sol:
```diff
@@ -0,0 +1,887 @@
+pragma solidity ^0.5.13;
```
Can we add a source of this file in the comment ?

karlb replied:
Done, see https://github.com/celo-org/celo-monorepo/pull/10240/commits/67038eeecbb8e3963f813a73fff1346d114327a3

karlb committed:
Add comment about origin of B12.sol

requested in https://github.com/celo-org/celo-monorepo/pull/10240#discussion_r1202392115

karlb committed:
Split BLS12_377_BASE into two ints

Constants of non-value type not yet implemented in solc, so we have to
create the Fp instance on every use, instead of using one global
instance.

karlb committed:
Rely on prettier for formatting of sol files

Since we ensure that all sol files are formatted by prettier, using
solhint for formatting checks provides no benefit and can cause style
conflicts. Just assume the prettier will format correctly and turn off
solhint's formatting checks.

montera82 reviewed (approved):
Welldone!
While i don't know much about cryptography, i looked through the upstream codebase where  this PR reffered to, and its in line from what i could see. Also the outstanding comments have been addressed

martinvol committed:
Merge branch 'master' into karlb/precompile-wrappers
